dask_searchcv/adaptive.py

```diff
@@ -0,0 +1,475 @@
+               "This model is %0.2f trained for this bracket")
+        logger.info(msg, k, s, i, percent=iter * 100.0 / max_iter)
+        if not dry_run:
+            model.partial_fit(X, y, **fit_kwargs)
+    fit_time = default_timer() - start_time
+    msg = ("Training model %s for %d partial_fit calls took %0.2f seconds")
+    logger.info(msg, k, max_iter, fit_time)
+
+    start_time = default_timer()
+    score = scorer(model, X_val, y_val) if not dry_run else np.random.rand()
+    score_time = default_timer() - start_time
+    msg = "Fitting model {k} took {fit} seconds and scoring took {score} seconds"
+    logger.info(msg.format(k=k, fit=fit_time, score=score_time))
+    return score, {'score_time': score_time, 'fit_time': fit_time}
+
+
+def _top_k(choose_from, eval_with, k=1):
+    assert k > 0, "k must be positive"
+    assert set(choose_from.keys()) == set(eval_with.keys())
+
+    keys = list(choose_from.keys())
+    assert k <= len(keys)
+
+    eval_with = [eval_with[key] for key in keys]
+    idx = np.argsort(eval_with)
+
+    return {keys[i]: choose_from[keys[i]] for i in idx[-k:]}
+
+
+def _random_choice(params, n=1):
+    configs = ParameterSampler(params, n)
+    return configs
+
+
+def rec_type(d):
+    if isinstance(d, (float, int, str, np.ndarray, da.Array)):
+        return type(d)
+    if isinstance(d, list):
+        return [rec_type(di) for di in d]
+    if isinstance(d, dict):
+        return {k: rec_type(v) for k, v in d.items()}
+    return type(d)
+
+
+def _successive_halving(params, model, data, n=None, r=None, s=None,
+                        shared=None, eta=3, _prefix='', dry_run=False,
+                        scorer=None,
+                        **fit_kwargs):
+    if any(x is None for x in [n, r, s, model, scorer]):
+        raise ValueError('n, r, and s are required')
+
+    if shared:
+        best = {k: shared[k] for k in ['val_score', 'config']}
+    else:
+        best = {'val_score': -np.inf, 'config': {}}
+
+    ids = [_prefix + '-' + str(i) for i in range(n)]
+    params = _random_choice(params, n=len(ids))
+    params = {k: param for k, param in zip(ids, params)}
+    models = {}
+    final_val_scores = {k: -np.inf for k in ids}
+    for k, param in params.items():
+        models[k] = clone(model).set_params(**param)
+
+    history = []
+    iters = 0
+    times = []
+    for i in range(s + 1):
+        n_i = math.floor(n * eta**-i)
+        r_i = r * eta**i
+        iters += r_i
+        msg = ('Training %d models for %d iterations during iteration %d '
+               'of bracket=%d')
+        logger.info(msg, n_i, r_i, i, s)
+
+        results = {k: dask.delayed(_train)(model, data, max_iter=r_i,
+                                           s=s, i=i, k=k, dry_run=dry_run,
+                                           scorer=scorer, **fit_kwargs)
+                   for k, model in models.items()}
+
+        results = {k: v.compute() for k, v in results.items()}
```

Comment:
We'll always want to call `dask.compute` on as much as possible at once.  This allows computations to share intermediate results.  You can call `dask.compute` on a dictionary and it will do the right thing:

```python
In [1]: import dask
da
In [2]: def inc(x):
   ...:     return x + 1
   ...: 

In [3]: d = {1: dask.delayed(inc)(1), 2: dask.delayed(inc)(2)}

In [4]: dask.compute(d)
Out[4]: ({1: 2, 2: 3},)
```

Reply:
I think I'm running into a bug here, at least when I use the distributed client. My test inconsistently pass/fail when I do

``` python
d = {k: dask.delayed(v) for k, v in ...}
dask.compute(d)[0]
```

in at least one of the loops (one in `_successive_halving`, one in `fit`).

They do consistently pass when I do `{k: v.compute() for k, v in ...}`, and the example notebook works too.

I'll try to construct a MWE tomorrow.

Reply:
If it is easy for you to provide a traceback then I recommend doing so.

Reply:
Here's the traceback from the notebook example: https://gist.github.com/stsievert/4ce7fa9ae0fed7e166b42f9d6a664a2e

I get this when I include `client.compute(delayed_results)` instead of `{k: v.compute() for k, v in delayed_results.items()}`. I only do this for the inner-most loop, in hopes it may be more interpretable. I also get an error when I do this with both for-loops, but have not verified the tracebacks are the same. This happens when I run `client = Client()`, and the error is thrown in the notebook without any intervention on my part.

Let me know if you what you can pull from that, or if you'd like a MWE.

Reply:
This error happens inconsistently too. I've run `Hyperband.fit` 4 times now, and it's passed once (with no changes to source or notebook in between).